Add `CustomLogger` to prevent IDE lint warnings when calling `log.trace`.

TizzySaurus opened the pull request:
Closes bot#1830.

Implements the proposed `CustomLogger` class to fix the lint warnings.

Xithrius commented:
[here](https://github.com/python-discord/bot/blob/1adae82ebc32fc39636f4ca2d20ec2e2b3c3fbd0/bot/log.py#L17-L30) is the `CustomLogger` class, so anyone reviewing this doesn't have to scroll to find it.

TizzySaurus commented:
> [Here](https://github.com/python-discord/bot/blob/1adae82ebc32fc39636f4ca2d20ec2e2b3c3fbd0/bot/log.py#L17-L30) is the `CustomLogger` class, so anyone reviewing this doesn't have to scroll to find it.

There's also the [`get_logger` util function](https://github.com/python-discord/bot/blob/1adae82ebc32fc39636f4ca2d20ec2e2b3c3fbd0/bot/utils/logging.py#L7-L9) (entire file is new)

All other changes are removing the old monkeypatch `trace` or updating imports from
```py
import logging

log = logging.getLogger(__name__)
```
to
```py
from bot.utils.logging import get_logger

log = get_logger(__name__)
```

kosayoda commented on bot/log.py:
```diff
@@ -14,11 +14,27 @@
 TRACE_LEVEL = 5
 
 
+class CustomLogger(Logger):
```
The logger should inherit [.getLoggerClass()](https://docs.python.org/3/library/logging.html?highlight=logging#logging.getLoggerClass) to play nice with other loggers.

TizzySaurus replied:
Changing this to `class CustomLogger(logging.getLoggerClass())` results in PyCharm giving a warning stating `Unresolved attribute reference 'isEnabledFor' for class 'CustomLogger'` and `Unresolved attribute reference '_log' for class 'CustomLogger'` (same happens for `log` instead of `_log`).

Even [`cast`](https://docs.python.org/3/library/typing.html#typing.cast)ing doesn't fix this (`cast(Logger, logging.getLoggerClass())`)

kosayoda replied:
> Unresolved attribute reference 'isEnabledFor' for class 'CustomLogger'

This seems to be a PyCharm bug, as both `mypy` and `pyright` correctly infers the correct type.

However, `mypy` (but not `pyright`) returns a different error relating to dynamic base classes. The workaround for that is
```python

if typing.TYPE_CHECKING:
    LoggerClass = Logger
else:
    LoggerClass = logging.getLoggerClass()

class CustomLogger(LoggerClass):
```
A nice explanation of why casting and even type variables will not work for this case can be found here: https://github.com/python/mypy/issues/11050
An even longer read (with a guido appearance) can be found here: https://github.com/python/mypy/issues/2477

TizzySaurus replied:
Committed as part of [ca2d4a5](https://github.com/python-discord/bot/pull/1831/commits/1adae82ebc32fc39636f4ca2d20ec2e2b3c3fbd0)

kosayoda commented on bot/log.py:
```diff
@@ -14,11 +14,27 @@
+
+        logger.trace("Houston, we have an %s", "interesting problem", exc_info=1)
+        """
+        if self.isEnabledFor(TRACE_LEVEL):
+            self._log(TRACE_LEVEL, msg, args, **kwargs)
```
It's better to use the non-internal `.log()` method to do the work for us, it essentially does [what you're doing here but extra](https://github.com/python/cpython/blob/778b07565e38cc94aa90318eb47b9cd09716756a/Lib/logging/__init__.py#L1536).

TizzySaurus replied:
So to clarify, this would become `self.log(...)`?

kosayoda replied:
Yep, and you pass `*args` instead of `args`

TizzySaurus replied:
Committed as part of [ca2d4a5](https://github.com/python-discord/bot/pull/1831/commits/1adae82ebc32fc39636f4ca2d20ec2e2b3c3fbd0)

kosayoda reviewed (dismissed):
Nice, I've personally wanted this for a while 👍🏻 

Missing log migration: `bot/exts/recruitment/talentpool/_cog.py`

TizzySaurus commented:
NB: Linting & Test currently failing due to [maintenance on the coveralls API](https://www.moesif.com/public-apis/status/coveralls). Will rerun checks later :+1:

TizzySaurus commented:
NB: The `getLogger` in  `bot.utils.members` will also need to be migrated once #1837 has been merged.

wookie184 commented on bot/constants.py:
```diff
@@ -25,7 +24,7 @@
     pass
 
-log = logging.getLogger(__name__)
+log = logging.getLogger(__name__)  # use logging.getLogger (not bot.utils.logging.get_logger) to prevent circular import
```
I'm not sure this is the best option here, there was some discussion on discord on this starting here https://discord.com/channels/267624335836053506/635950537262759947/894235145362616390, and it seems the conclusion was that the logs here don't work anyway because this is run before logging is set up.

TizzySaurus replied:
This has been resolved with the little re-write :+1:

wookie184 commented on bot/utils/services.py:
```diff
@@ -6,7 +6,7 @@
 from bot.constants import URLs
 
-log = logging.getLogger(__name__)
+log = logging.getLogger(__name__)  # use logging.getLogger (not bot.utils.logging.get_logger) to prevent circular import
```
Similar to other comment, although I feel we should be able to use the logger here, and using `get_logger` for it would be nice, although i'm not sure what the best way forwards with that is. This problem only seems to happen because this file is imported in the `utils/__init__.py`

TizzySaurus replied:
This has been resolved with the little re-write :+1:

wookie184 reviewed (commented):
I'm not sure using `log = logging.getLogger(__name__)` is the best solution to the circular import problem, although i'm not sure what is either...

Other than that this looks good.

TizzySaurus commented:
Have fixed the root issue of the circular imports, so now all files should be using `bot.log.get_logger` instead of `logging.getLogger`.

Also added any missing migrations from PRs made after I started this and squashed, fixed the conflicts (all 25), and did a commit for making newlines after imports consistent (part of PyCharm's "Optimize Imports" feature).

This should now be fully read for review @kosayoda @wookie184 :+1:

wookie184 commented on bot/constants.py:
```diff
@@ -123,11 +119,11 @@ def check_required_keys(keys):
                     raise KeyError(key)
         except KeyError:
-            log.critical(
+            raise (
```
Need an exception class here, `KeyError` would probably be good, So `raise KeyError("message")`. That or we could print and then just `raise`, maybe that would be better as it would keep the context of where the KeyError was raised? Can't remember exactly how these things work

wookie184 reviewed (changes requested):
Two small things, other than that it looks good.

TizzySaurus commented on bot/constants.py:
```diff
@@ -186,8 +182,8 @@ def __getattr__(cls, name):
             )
-            # Only an INFO log since this can be caught through `hasattr` or `getattr`.
-            log.info(f"Tried accessing configuration variable at `{dotted_path}`, but it could not be found.")
+            # Only an print since this can be caught through `hasattr` or `getattr`.
+            print(f"Tried accessing configuration variable at `{dotted_path}`, but it could not be found.")
```
Removed :+1:
